driver/docker-container/driver.go

```diff
@@ -196,6 +200,126 @@ func (d *Driver) copyLogs(ctx context.Context, l progress.SubLogger) error {
+		}
+		dstInfo.Path = linkTarget
+		dstStat, err = d.DockerAPI.ContainerStatPath(ctx, d.Name, linkTarget)
+	}
+
+	// Validate the destination path.
+	if err := command.ValidateOutputPathFileMode(dstStat.Mode); err != nil {
+		return errors.Wrapf(err, `destination "%s:%s" must be a directory or a regular file`, d.Name, dstPath)
+	}
+
+	// Ignore any error and assume that the parent directory of the destination
+	// path exists, in which case the copy may still succeed. If there is any
+	// type of conflict (e.g., non-directory overwriting an existing directory
+	// or vice versa) the extraction will fail. If the destination simply did
+	// not exist, but the parent directory does, the extraction will still
+	// succeed.
+	dstInfo.Exists, dstInfo.IsDir = true, dstStat.Mode.IsDir()
+
+	// Prepare source copy info.
+	srcInfo, err := dockerarchive.CopyInfoSourcePath(srcPath, true)
+	if err != nil {
+		return err
+	}
+
+	srcArchive, err := dockerarchive.TarResource(srcInfo)
+	if err != nil {
+		return err
+	}
+	defer srcArchive.Close()
+
+	// With the stat info about the local source as well as the
+	// destination, we have enough information to know whether we need to
+	// alter the archive that we upload so that when the server extracts
+	// it to the specified directory in the container we get the desired
+	// copy behavior.
+
+	// See comments in the implementation of `dockerarchive.PrepareArchiveCopy`
+	// for exactly what goes into deciding how and whether the source
+	// archive needs to be altered for the correct copy behavior when it is
+	// extracted. This function also infers from the source and destination
+	// info which directory to extract to, which may be the parent of the
+	// destination that the user specified.
+	dstDir, preparedArchive, err := dockerarchive.PrepareArchiveCopy(srcArchive, srcInfo, dstInfo)
+	if err != nil {
+		return err
+	}
+	defer preparedArchive.Close()
+
+	return d.DockerAPI.CopyToContainer(ctx, d.Name, dstDir, preparedArchive, dockertypes.CopyToContainerOptions{
+		AllowOverwriteDirWithFile: false,
+	})
+}
+
+func resolveLocalPath(localPath string) (absPath string, err error) {
+	if absPath, err = filepath.Abs(localPath); err != nil {
+		return
+	}
+	return dockerarchive.PreserveTrailingDotOrSeparator(absPath, localPath, filepath.Separator), nil
+}
+
+func (d *Driver) copyRegCertsToContainer(ctx context.Context, bkconfig string) error {
+	// Load BuildKit config.
+	bcfg, err := buildkitconfig.LoadFile(bkconfig)
+	if err != nil {
+		return err
+	}
+
+	// Create a temp folder and copy all certs to it with copyFileParents
+	// to keep the exact same folder structure as defined in BuildKit config.
+	// It's required to do that here because we cannot copy a file in a folder
+	// that does not exist in the container with the CopyToContainer API.
+	tmpRegCerts, err := os.MkdirTemp("", "basedir-certs")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpRegCerts)
```

Comment:
Is there a better way to do this with Docker API (cc @tonistiigi @thaJeztah)? Would also be better to use a tmpfs for security.

Reply:
I think we should try to rewrite the cert paths in config and reencode it so the certs always point to `/etc/buildkit/certs` or something. The only issue would be making sure that if buildkit config gets updated it still works. Maybe parse it to https://pkg.go.dev/github.com/pelletier/go-toml#Tree

Reply:
> I think we should try to rewrite the cert paths in config and reencode it so the certs always point to `/etc/buildkit/certs` or something.

Yes that could do it. I wonder if an extension of the root ca and key pairs fields in buildkitd config to allow base64 encoded content would not be a better alternative.